ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -0,0 +1,176 @@
+recordProgress file shift InProgress{..} = case HMap.alterF alter file current of
+    ((prev, new), m') ->
+        let todo' = if isJust prev then todo else todo + 1
+            done' = if new == 0 then done+1 else done
+        in InProgress todo' done' m'
+  where
+    alter x = let x' = maybe (shift 0) shift x in ((x,x'), Just x')
+
+-- | A 'ProgressReporting' that enqueues Begin and End notifications in a new
+--   thread, with a grace period (nothing will be sent if 'KickCompleted' arrives
+--   before the end of the grace period).
+delayedProgressReporting
+  :: Seconds  -- ^ Grace period before starting
+  -> Seconds  -- ^ sampling delay
+  -> Maybe (LSP.LanguageContextEnv c)
+  -> ProgressReportingStyle
+  -> IO ProgressReporting
+delayedProgressReporting before after lspEnv optProgressStyle = do
+    inProgressVar <- newVar $ InProgress 0 0 mempty
+    progressState <- newVar NotStarted
+    let progressUpdate event = modifyVar_ progressState $ updateState (mRunLspT lspEnv $ lspShakeProgress inProgressVar) (Event event)
+        progressStop   = modifyVar_ progressState $ updateState (mRunLspT lspEnv $ lspShakeProgress inProgressVar) StopProgress
+        inProgress :: NormalizedFilePath -> Action a -> Action a
+        inProgress = withProgressVar inProgressVar
+    return ProgressReporting{..}
+    where
+        lspShakeProgress inProgress = do
+            -- first sleep a bit, so we only show progress messages if it's going to take
+            -- a "noticable amount of time" (we often expect a thread kill to arrive before the sleep finishes)
+            liftIO $ sleep before
+            u <- ProgressTextToken . T.pack . show . hashUnique <$> liftIO newUnique
+
+            b <- liftIO newBarrier
+            void $ LSP.sendRequest LSP.SWindowWorkDoneProgressCreate
+                LSP.WorkDoneProgressCreateParams { _token = u } $ liftIO . signalBarrier b
+            ready <- liftIO $ waitBarrier b
+
+            for_ ready $ const $ bracket_ (start u) (stop u) (loop u 0)
+            where
+                start id = LSP.sendNotification LSP.SProgress $
+                    LSP.ProgressParams
+                        { _token = id
+                        , _value = LSP.Begin $ WorkDoneProgressBeginParams
+                          { _title = "Processing"
+                          , _cancellable = Nothing
+                          , _message = Nothing
+                          , _percentage = Nothing
+                          }
+                        }
+                stop id = LSP.sendNotification LSP.SProgress
+                    LSP.ProgressParams
+                        { _token = id
+                        , _value = LSP.End WorkDoneProgressEndParams
+                          { _message = Nothing
+                          }
+                        }
+                loop id prev = do
+                    InProgress{..} <- liftIO $ readVar inProgress
+                    if todo == 0 then loop id 0 else do
```

Comment:
If we hit loop id 0, then we spin at 100% CPU. That path seems like it should still have a sleep on it.

Reply:
Good catch!